google/cloud/internal/error_metadata.h

```diff
@@ -0,0 +1,158 @@
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+/**
+ * A (relatively) lightweight data structure to pass error metadata across
+ * implementation functions.
+ *
+ * Sometimes we want to provide additional context about errors. The original
+ * motivation is credential file parsing. These files can be fairly complex,
+ * and parsing requires many functions that only need the *contents* of the
+ * file to parse it, but may want to show the filename, the start of the
+ * parsing call tree, and maybe some key intermediate callers.
+ *
+ * @par Example
+ * @code
+ * namespace google::cloud::internal {
+ * StatusOr<Foo> ParseFooFile(std::string filename) {
+ *   ErrorContext error_context{
+ *     {"filename", filename},
+ *     {"origin", __func__},
+ *   };
+ *   std::ifstream is(filename);
+ *   auto contents = std::string{std::istreambuf_iterator<char>{is.buf()},{ }};
+ *   if (is.bad()) {
+ *     return Status(
+ *        StatusCode::kInvalidArgument,
+ *        Format("cannot read file", error_context));
+ *   }
+ *   return ParseFooFileContents(
+ *       std::move(contents), std::move(error_context));
+ * }
+ *
+ * StatusOr<Foo> ParseFooFileContents(
+ *     std::string contents, ErrorContext error_context) {
+ *   // Do some stuff
+ *   if (has_bar()) return ParseFooFileContentsWithBar(
+ *       .., std::move(error_context));
+ *   // do more stuff
+ *   if (bad()) return Status(
+ *       StatusCode::kInvalidArgument,
+ *       Format("badness parsing thingamajig", error_context));
+ *   // all good
+ *   return Foo{...};
+ * }
+ * @endcode
+ */
+class ErrorContext {
+ public:
+  using Container =
+      std::vector<std::pair<absl::string_view, absl::string_view>>;
+
+  ErrorContext() = default;
+  explicit ErrorContext(Container m) : metadata_(std::move(m)) {}
+
+  ErrorContext(ErrorContext const&) = delete;
+  ErrorContext& operator=(ErrorContext const&) = delete;
+  ErrorContext(ErrorContext&&) = default;
+  ErrorContext& operator=(ErrorContext&&) = default;
+
+  template <typename... A>
+  Container::reference emplace_back(A&&... a) {
+    return metadata_.emplace_back(std::forward<A>(a)...);
+  }
+
+  void push_back(Container::value_type p) {
+    return metadata_.push_back(std::move(p));
+  }
+
+  bool empty() const { return metadata_.empty(); }
+
+  Container::const_iterator begin() const { return metadata_.begin(); }
+
+  Container::const_iterator end() const { return metadata_.end(); }
+
+ private:
+  Container metadata_;
+};
+
+std::string Format(absl::string_view message, ErrorContext const& context);
+
+/**
+ * Stores an error context for later use.
+ *
+ * In most cases an error context is used in a single call tree, and therefore
+ * using `absl::string_view` gives a good performance tradeoff.  Sometimes one
+ * needs to capture the error context for use after the function returns.  In
```

Comment:
It's a little more insidious than "after the function returns", methinks.  `ErrorContext` is a non-owning object, and its lifetime needs to be contained within the lifetime of its ctor argument, so its doc block probably deserves a big warning to that effect (or something like ABSL_ATTRIBUTE_LIFETIME_BOUND perhaps).

Reply:
I think I will switch this to have simpler lifetime semantics.  The use-cases I have in mind are not in the critical path and are not worth the extra thinking required to avoid problems.

In the future we may care about avoiding allocations for a bag of attributes that is unlikely to be used.